### PR TITLE
Core: Ability to run in sandboxed iframe, (#1090)

### PR DESCRIPTION
**Pull request by Gaurav0**
Do not directly refer to window.sessionStorage (#1090)

### CODE REVIEW

**Comment by Gaurav0** [comment]
@platinumazure I have added the unit test as requested. Some changes to the test harness were required.

**Comment by platinumazure** [comment]
@trentmwillis @leobalter Can one of you please review this? I inadvertently requested a change which resulted in an overhaul of the test harness, is that change okay?

File: test/sandbox-inside.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html>
+	<head>
+		<meta charset="UTF-8">
+		<title>QUnit Only Test Suite</title>
```
Comment (trentmwillis) [inline test/sandbox-inside.html]: Nitpick: Let's update the title here

File: src/globals.js [hunk 2]
```diff
@@ -7,4 +7,19 @@ export const clearTimeout = global.clearTimeout;
 
 export const document = window && window.document;
 export const navigator = window && window.navigator;
-export const sessionStorage = window && window.sessionStorage;
+
+let defined = {
+	sessionStorage: ( function() {
```
Comment (trentmwillis) [inline src/globals.js]: Let's just export the result directly:
```js
export const sessionStorage = ( function() { ... }() );
```
No need to have an intermediate value.

File: package.json [hunk 3]
```diff
@@ -40,6 +40,7 @@
     "grunt": "1.0.1",
     "grunt-cli": "1.2.0",
     "grunt-concurrent": "2.3.1",
+    "grunt-contrib-connect": "^1.0.2",
```
Comment (leobalter) [inline package.json]: conventionally, we lock the package versions. Can you remove the `^` here, please?

File: src/globals.js [hunk 4]
```diff
@@ -7,4 +7,19 @@ export const clearTimeout = global.clearTimeout;
...
+		var x = "qunit-test-string";
+		try {
+			sessionStorage.setItem( x, x );
+			sessionStorage.removeItem( x );
+			return sessionStorage;
```
Comment (leobalter) [inline src/globals.js]: we could use something better than a try catch, something like the `in` operator as `return 'setItem' in sessionStorage && 'removeItem' in sessionStorage && sessionStorage`

This way we also don't need to catch any false alarms if sessionStorage doesn't work on storing or removing an item. All we want is to confirm the API is available.

File: test/sandbox.html [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Comment (leobalter) [inline test/sandbox.html]: sandbox and sandbox-inside doesn't play well with the other html files in this folder. It's repeating another test and the intent is implicit only. After this PR one would look and ask why this is here. The same can be asked when reading the Gruntfile.

naming is hard, be we should try again w/ this one. Maybe sessionstorage-through-iframe.html is long but good enough, I'm bad with names too.

File: Gruntfile.js [hunk 6]
```diff
@@ -10,6 +10,7 @@ module.exports = function( grunt ) {
 
 // Load grunt tasks from NPM packages
 require( "load-grunt-tasks" )( grunt );
+grunt.loadNpmTasks( "grunt-contrib-connect" );
```
Comment (leobalter) [inline Gruntfile.js]: I believe this can be discarded as the line above should handle this.

File: src/globals.js (hunk 4 above)
Comment (Gaurav0) [inline src/globals.js]: Unfortunately it blows up in the sandbox if we do that.

File: Gruntfile.js (hunk 6 above)
Comment (Gaurav0) [inline Gruntfile.js]: Thanks!

File: package.json (hunk 3 above)
Comment (Gaurav0) [inline package.json]: Done.

File: src/globals.js (hunk 2 above)
Comment (Gaurav0) [inline src/globals.js]: Yes, makes sense.

File: test/sandbox-inside.html (hunk 1 above)
Comment (Gaurav0) [inline test/sandbox-inside.html]: Done.

File: test/sandbox.html (hunk 5 above)
Comment (Gaurav0) [inline test/sandbox.html]: I'm open to suggestions, but don't want a name specific to sessionStorage issue, this test catches any error when loading QUnit through an iframe.

Comment (trentmwillis) [inline test/sandbox.html]: Can we do `sanboxed-iframe` and `sandboxed-iframe-contents`? I think that is a good balance between overly specific and generic

Comment (Gaurav0) [inline test/sandbox.html]: Yes!

**Comment by trentmwillis** [comment]
Looks good to me! @leobalter / @platinumazure any other concerns?

**Comment by platinumazure** [comment]
Oh, actually, one suggestion. Do we need to load a test file at all actually? Maybe the iframe content file could just use an inline `<script>` tag and just do a quick inline test with an `assert.ok(true)` or something?

File: src/globals.js [hunk 7]
```diff
@@ -7,4 +7,14 @@ export const clearTimeout = global.clearTimeout;
...
-export const sessionStorage = window && window.sessionStorage;
+
+export const sessionStorage = ( function() {
+	var x = "qunit-test-string";
+	try {
+		sessionStorage.setItem( x, x );
+		sessionStorage.removeItem( x );
+		return sessionStorage;
+	} catch ( e ) {
+		return undefined;
```
Comment (leobalter) [inline src/globals.js]: I don't like this try catch block, but that's something I can investigate later how to improve it.

Comment (trentmwillis) [inline src/globals.js]: Agreed, but given the nature of the error I think it unavoidable.

File: src/globals.js [hunk 8]
```diff
@@ -7,4 +7,14 @@ export const clearTimeout = global.clearTimeout;
 
 export const document = window && window.document;
 export const navigator = window && window.navigator;
-export const sessionStorage = window && window.sessionStorage;
+
+export const sessionStorage = ( function() {
```
Comment (stefanpenner) [inline src/globals.js]: this `const sessionStorage` shadows the global one, forcing the `sessionStorage` in the function scope (line 14) to always be undefined.

Comment (stefanpenner) [inline src/globals.js]: one approach to fixing: https://github.com/qunitjs/qunit/pull/1112

**Pull request merged**